features: install, uninstall, search, local_plugins

bucko13 opened the pull request:
Adds a lot of new functionality to bpanel-cli tools. Can test by running `node ./dist/program.js --help` to see available commands and expected args

- `i` or `install` to add a new plugin to config.js
- `u` or `u` to uninstall plugin from config.js
- `s` or `search` to search npm for bpanel plugins
-  `create` defaults to creating a new plugin in the `[BPANEL_PREFIX]/local_plugins` directory
- can optionally pass a `--prefix` option to override this default or still pass absolute path as before.

bucko13 committed:
default install into local_plugins, use bcfg, bsert, blgr utilities

bucko13 committed:
add support for install plugin script

bucko13 committed:
add support for uninstalling plugins

bucko13 committed:
add search command

bucko13 committed:
update readme

tynes commented on lib/program.js:
```diff
@@ -1,16 +1,79 @@
+(async function() {
+  await logger.open();
+
+  program
```
does `bcfg` have cli tools built in? I remember seeing something with how nodar used it for the handshake address generator

tynes replied:
This would allow for us to eliminate a dependency. It may be good to learn how to use our tool to build cli tools, but also there could be more important things to do

bucko13 replied:
It doesn't really have CLI tools per se, just the ability to parse options from the command line. It would be possible to do this with bcfg but it would require building out some of the scaffolding that `commander` and `inquirer` provide out of the box, e.g. options parsing, help menu, and the convenience methods that allow the code to be a little more declarative. I'm not against using bcfg, but for what it's worth, bcfg was in the process of being separated out when this was first built, so migrating over to it would require a decent amount of refactoring.

tynes replied:
Ok, makes sense, this looks good to me

bucko13 committed:
useability improvements and new list command

bucko13 commented:
@tynes Added one more commit. Needed to add parser config to uses of `prettier.format` to avoid a warning message, cleaned up error reporting, and added a new command to list all currently installed plugins.

bucko13 commented:
I'm thinking it would make sense for this to be a minor, or even a major 1.0 release after this is merged in. Nothing really backwards incompatible, but this makes the tool much more useful in my opinion and a little more deserving a proper major version.

bucko13 committed:
support for installing multiple plugins at once and other usability improvements

tynes commented on lib/create.js:
```diff
@@ -122,26 +137,27 @@ export default async () => {
     if (dependencies) {
-      console.log(chalk.green('Installing plugin dependencies from npm'));
+      logger.info(chalk.green('Installing plugin dependencies from npm'));
       execSync(`npm install --save ${dependencies.replace(/,/g, '')}`, {
```
Would it be a good idea validate dependencies before just passing it into `npm install`?

bucko13 replied:
I think this is along the same lines as the other comment. I don't think there's any more need for us to validate dependencies since that's basically what `npm install` will do and this is really just a wrapper around a user running `npm install` themselves. I can't see a circumstance where we need to stop someone from injecting arbitrary code execution on themselves that they wouldn't couldn't otherwise do from including it in a normal `npm install` command. There is also some validation that already happens in `questions.js` where this is set. I can add an extra check using npm's `validate` package though to make sure the names don't have any typos.

tynes commented on lib/helpers.js:
```diff
@@ -122,3 +127,15 @@ export const setupModule = targetComponent => module => {
   return moduleTemplate;
 };
+
+export function checkExistence(name, list = [], local = false) {
```
This function feels a bit weird to me due to it being polymorphic, it seems like its designed to be in a `try/catch` where it would throw the error message in the object?

bucko13 replied:
Yeah, that's originally what I had it as actually, but because I was using it in `install` and `uninstall` it felt DRYer to abstract it like this and it couldn't throw since you're looking for a different outcome in each circumstance.

tynes commented on lib/install.js:
```diff
@@ -0,0 +1,82 @@
+
+      const exists = checkExistence(names[i], configList, cmd.local);
+
+      if (exists) {
```
A future person refactoring may introduce an error (wouldn't be fatal in this case) if `exists` returns as `false` and then `exists.message` is referenced. It seems like a leaky abstraction, as you need to know the logic happening inside of `checkExistence` to be able to refactor this code

bucko13 replied:
Yeah, I think that's fair. I'll see if I can come up with something to plug it up a bit.

tynes commented on lib/program.js:
```diff
@@ -1,16 +1,86 @@
+      'Search for available plugins to install'
+    )
+    .action(() => {
+      const output = execSync(`npm search bpanel ${process.argv.slice(3).join(' ')}`, {
```
A user could execute arbitrary code here, I don't think its the hugest deal but its still something that we should think about. I could imagine a person giving a malicious command to copy/paste here, it could feel safe because its a `bpanel` command

bucko13 replied:
My feeling is that this isn't any different from a user entering any other commands. i.e. you would only be running this on your own system. It doesn't allow for an outside user to execute arbitrary code on your own system, just yourself. This doesn't feel much different to me from `npm install` or `npm search` allowing someone to execute arbitrary code.

bucko13 replied:
For example, this is no less a risk than if someone gave an uninformed user the command `npm search ; rm -rf ~` to copy and paste into their terminal.

tynes commented:
One gotcha that I found was in the multiselect parts, where you have to use spacebar to toggle selection. It wasn't particularly clear that is how you do it

tynes commented on lib/helpers.js:
```diff
@@ -49,49 +49,54 @@ export const setupDeps = async (deps_, pluginRoot) => {
     import: format(`import modules from './plugins';`),
     arr: format(
-      `const plugins = Object.keys(modules).map(name => modules[name]);`
+      `const plugins = Object.keys(modules).map(name => modules[name]);`,
```
This is equal to `Object.values(modules)`

bucko13 replied:
good catch

bucko13 committed:
update install to only install for modules that exist either locally or on npm

bucko13 commented:
> One gotcha that I found was in the multiselect parts, where you have to use spacebar to toggle selection. It wasn't particularly clear that is how you do it

lol, yeah that still gets me too. I added a little comment to prompt to make more clear. 

Also just pushed up another update to install that will skip any plugins that don't exist locally or on npm (depending on installation type).

tynes commented on lib/install.js:
```diff
@@ -0,0 +1,101 @@
+function filterArgs() {
+  // remove node, path, and command args
+  const args = process.argv.slice(3);
+  let names = args.filter(name => name[0] !== '-');
```
What about `--`? I think that a regex would be more useful

bucko13 committed:
npmExists support redirects

tynes commented on lib/helpers.js:
```diff
@@ -122,3 +130,25 @@ export const setupModule = targetComponent => module => {
+  const base = 'https://www.npmjs.org/package/';
+  const { href: path } = new URL(packageName, base);
+  const response = await brq({ url: path, method: 'HEAD' });
+  if (200 < response.statusCode < 400) return true;
```
`200 <= statuscode < 400`

tynes commented on lib/install.js:
```diff
@@ -0,0 +1,101 @@
+}
+
+module.exports = (config, logger) => async (name, cmd) => {
+  let bpanelConfig, CONFIG_PATH;
```
Any reason these are declared at the top? Everything happens inside of the `try`

bucko13 replied:
good catch, that was left over from a previous iteration.

tynes commented on lib/uninstall.js:
```diff
@@ -0,0 +1,73 @@
+const { checkExistence } = require('./helpers');
+
+module.exports = (config, logger) => async (name, cmd) => {
+  let bpanelConfig, configPath;
```
`CONFIG_PATH` vs `configPath`

bucko13 replied:
Yep, removing this line now.

bucko13 committed:
cleanup

bucko13 committed:
fix statusCode check
